// greasemonkey/lib/xhr.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const api_1 = require("./api");
const bluebird = require("bluebird");
exports.GM_XHR = api_1.default.xmlhttpRequest;
function makeOptions(options) {
    let data = Object.assign({
        method: 'GET',
    }, options);
    if (data.xhrFields) {
        Object.assign(data, data.xhrFields);
    }
    return data;
}
exports.makeOptions = makeOptions;
function ajax(url, options) {
    options = makeOptions(options);
    options.url = url;
    if (options.data && typeof options.data == 'object' && !(options.data instanceof FormData)) {
        let formData = new FormData();
        for (let i in options.data) {
            formData.append(i, options.data[i]);
        }
        options.data = formData;
    }
    return new bluebird(function (resolve, reject) {
        exports.GM_XHR(Object.assign({}, options, {
            onload: function (res) {
                resolve(res);
            },
            onerror: reject,
        }));
    })
        .then(function (res) {
        if (options.dataType == 'json') {
            res.responseOriginal = res.response;
            res.response = JSON.parse(res.responseText);
        }
        res.responseType = res.responseType || options.dataType || null;
        if (typeof options.onload === 'function') {
            let ret = options.onload.call(this, res);
            if (typeof ret !== 'undefined') {
                return ret;
            }
        }
        return res;
    });
}
exports.ajax = ajax;
function AjaxError(res, libError = Error) {
    let err = new libError(res);
    err.res = res;
    return err;
}
exports.AjaxError = AjaxError;
exports.default = exports;
